### PR TITLE
osd: don't carry PGLSFilter between multiple ops in MOSDOp.

### PR DESCRIPTION
**Pull request by rzarzynski**
Currently result of executing multiple PG listing operations in single `MOSDOp` – if `PGLSFilter` is being used – depends on the order in the sequence. To exemplify:
```
  [ CEPH_OSD_OP_PGNLS, CEPH_OSD_OP_PGNLS_FILTER ]
```
may bring different result for `CEPH_OSD_OP_PGNLS` than
```
  [ CEPH_OSD_OP_PGNLS_FILTER, CEPH_OSD_OP_PGNLS ].
```
This happens because `..._PGNLS_FILTER` sets up `PGLSFilter` which is used also by plain `..._PGNLS`.

The commit changes that. After the filter is set-up separately for each interested operation in `MOSDOp`.

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -1082,14 +1081,14 @@ void PrimaryLogPG::do_pg_op(OpRequestRef op)
...
   string cname, mname;
-  PGLSFilter *filter = NULL;
   bufferlist filter_out;
 
   snapid_t snapid = m->get_snapid();
 
   vector<OSDOp> ops = m->ops;
 
   for (vector<OSDOp>::iterator p = ops.begin(); p != ops.end(); ++p) {
+    std::unique_ptr<PGLSFilter> filter;
```
Comment (rzarzynski) [inline src/osd/PrimaryLogPG.cc]: That's the key change in the commit. I'm not sure whether the carrying behavior is intended or not. This PR is just an excuse to ignite discussion.

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -837,47 +837,48 @@ bool PGLSPlainFilter::filter(const hobject_t &obj,
   return true;
 }
 
-bool PrimaryLogPG::pgls_filter(PGLSFilter *filter, hobject_t& sobj, bufferlist& outdata)
+bool PrimaryLogPG::pgls_filter(PGLSFilter& filter, hobject_t& sobj, bufferlist& outdata)
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: i'd suggest extract this change out of this commit.

Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: also, i think another step is to mark the methods of `filter()` , `get_xattr()` and `reject_empty_xattr()` const, to assure that they do not have side-effect on `PGLSFilter`. and then we can safely move local variable`filter` in `do_pg_op()` into the loop.

Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: Extracted the key change from the refactoring and fixed the `const`-markings for `PGLSFilter`. I think the new commits will be useful also in the Ceph Class PR to address  [the `TODO` in `pgls_filter()`](https://github.com/ceph/ceph/pull/29651/files#diff-8a0759a622fe67c09d3bf6e5002ab77aR165).

**Comment by rzarzynski** [comment]
@tchaikov: rebased after the `PGLSParentFilter`.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -1053,11 +1053,12 @@ void PrimaryLogPG::do_pg_op(OpRequestRef op)
 	result = -EINVAL;
 	break;
       }
-      if (filter) {
-	delete filter;
-	filter = NULL;
+
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: might want to remove this empty line.

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -1223,11 +1224,12 @@ void PrimaryLogPG::do_pg_op(OpRequestRef op)
...
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: ditto.

File: src/osd/PrimaryLogPG.cc (hunk 3 above)
Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: Removed.

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by dzafman** [comment]
@rzarzynski Should we create a tracker (referencing in the first commit) for backports?

**Comment by liewegas** [comment]
Not sure a backport is needed for this one?

**Comment by rzarzynski** [comment]
Nope, we don't need to backport these commits.

**Pull request merged**